### PR TITLE
[CPP Graph] MPT MHA support

### PR DESCRIPTION
**Pull request by DDEle**
## Type of Change: feature
API not changed

## Description
Add MHA fusion support for MPT including [ALiBi](https://arxiv.org/abs/2108.12409) support in fused attention operation.

## Expected Behavior & Potential Risk
N/A

## How has this PR been tested?
This PR is tested with CI vi GHA and local tests for accuracy and performance.

As a result, the model can generate reasonable text with MHA fusion enabled, and it preserves the identical behavior when disable the fusion (with `--memory-f16`). The fusion brings ~850% (12656.38ms => 1330.82ms) performance for first token and ~40% (73.29ms => 52.63ms) performance for next token inference.

In addition, it is tested that llama & gptj will preserve its behavior in this PR (q4j-int8-g128). 

<details>
<summary>Accuracy Tests Details:</summary>

Commands:
```bash
set -ex
#main branch
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=OFF -DCMAKE_BUILD_TYPE=Release && ninja run_mpt && numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 256 --memory-auto -p "$GIRL_PROMPT"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=OFF -DCMAKE_BUILD_TYPE=Release && ninja run_mpt && numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 256 --memory-auto -p "$(echo "$LUOYU_PROMPT" | cut -d' ' -f 1-800)"

#graph-mpt-mha-support
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=OFF -DCMAKE_BUILD_TYPE=Release && ninja run_mpt && numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 256 --memory-f16 -p "$GIRL_PROMPT"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=OFF -DCMAKE_BUILD_TYPE=Release && ninja run_mpt && numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 256 --memory-auto -p "$GIRL_PROMPT"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=OFF -DCMAKE_BUILD_TYPE=Release && ninja run_mpt && numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 256 --memory-f16 -p "$(echo "$LUOYU_PROMPT" | cut -d' ' -f 1-800)"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=OFF -DCMAKE_BUILD_TYPE=Release && ninja run_mpt && numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 256 --memory-auto -p "$(echo "$LUOYU_PROMPT" | cut -d' ' -f 1-800)"
```

Results:
- <details>
  <summary>
  main branch <code>numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 256 --memory-auto -p "$GIRL_PROMPT"</code>
  </summary>

  > Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun. But the only place she could think of that was interesting enough for her adventure-loving heart isâ¦ The beach! [end of text]
  </details>
- <details>
  <summary>
  main branch <code>numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 256 --memory-auto -p "$(echo <a href="https://pastebin.com/Qgs4X0VC">"$LUOYU_PROMPT"</a> | cut -d' ' -f 1-800)"</code>
  </summary>

  > ... If no spaghetti ever reaches all plates on the table (and dies) you win! [end of text]
  </details>
- <details>
  <summary>
  graph-mpt-mha-support <code>numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 256 --memory-f16 -p "$GIRL_PROMPT"</code>
  </summary>

  > Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun. But the only place she could think of that was interesting enough for her adventure-loving heart isâ¦ The beach! [end of text]
  </details>
- <details>
  <summary>
  graph-mpt-mha-support <code>numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 256 --memory-auto -p "$GIRL_PROMPT"</code>
  </summary>

  > Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun. But the only place she could ever see was her own house in Texas! [end of text]
  </details>
- <details>
  <summary>
  graph-mpt-mha-support <code>numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 256 --memory-f16 -p "$(echo <a href="https://pastebin.com/Qgs4X0VC">"$LUOYU_PROMPT"</a> | cut -d' ' -f 1-800)"</code>
  </summary>

  > ... If no spaghetti ever reaches all plates on the table (and dies) you win! [end of text]
  </details>
- <details>
  <summary>
  graph-mpt-mha-support <code>numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 256 --memory-auto -p "$(echo <a href="https://pastebin.com/Qgs4X0VC">"$LUOYU_PROMPT"</a> | cut -d' ' -f 1-800)"</code>
  </summary>

  > ... So this makes me wonder: what should I call my game? How about 'Pasta Evolution'? The evolution idea and all-powerful aspect fits in quite well there, but that's not much of an RPG... If anything really stands out is how every player has their own plate to control (and a team/side). So now you can see why this concept wasn't as good. It seems way too basic for my liking; I wanted the game itself was more interesting than just having fun killing pasta on plates and shooting sauce, but of course it's not so simple in real life is that? Anyway! [end of text]
  </details>
</details>

<details>
<summary>Performance Tests Details:</summary>

Commands:
```bash
set -ex
#main branch
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=ON -DCMAKE_BUILD_TYPE=Release && ninja run_mpt && env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 3 --memory-auto -p "$(echo "$LUOYU_PROMPT" | cut -d' ' -f 1-800)"

#graph-mpt-mha-support
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=ON -DCMAKE_BUILD_TYPE=Release && ninja run_mpt && env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 48 -b 2048 -c 2048 -n 3 --memory-auto -p "$(echo "$LUOYU_PROMPT" | cut -d' ' -f 1-800)"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=ON -DCMAKE_BUILD_TYPE=Release && ninja run_mpt && env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 3 --memory-auto -p "$(echo "$LUOYU_PROMPT" | cut -d' ' -f 1-800)"
rm -rf bin && cmake .. -GNinja -DNE_BUILD_TESTS=ON -DNE_PROFILING=ON -DCMAKE_BUILD_TYPE=Release && ninja run_mpt && env ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 3 --memory-f16 -p "$(echo "$LUOYU_PROMPT" | cut -d' ' -f 1-800)"
```

Results:
- <details>
  <summary>
  main branch <code>ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 3 --memory-auto -p "$(echo <a href="https://pastebin.com/Qgs4X0VC">"$LUOYU_PROMPT"</a> | cut -d' ' -f 1-800)"</code>
  </summary>

  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =  23.541 ms
  perf_total_per_op_us[                     MUL] =  25.601 ms
  perf_total_per_op_us[                  REPEAT] = 136.490 ms
  perf_total_per_op_us[                    GELU] = 188.008 ms
  perf_total_per_op_us[                    NORM] =  27.599 ms
  perf_total_per_op_us[                 MUL_MAT] = 2595.437 ms
  perf_total_per_op_us[                   SCALE] = 1141.293 ms
  perf_total_per_op_us[                     CPY] = 138.489 ms
  perf_total_per_op_us[                 RESHAPE] =   0.107 ms
  perf_total_per_op_us[                    VIEW] =   0.812 ms
  perf_total_per_op_us[                 PERMUTE] =   0.330 ms
  perf_total_per_op_us[               TRANSPOSE] =   0.110 ms
  perf_total_per_op_us[                GET_ROWS] =   5.255 ms
  perf_total_per_op_us[           DIAG_MASK_INF] = 368.837 ms
  perf_total_per_op_us[                SOFT_MAX] =  72.537 ms
  perf_total_per_op_us[                   ALIBI] = 7926.504 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.357 ms
  perf_total_per_op_us[                     MUL] =   0.384 ms
  perf_total_per_op_us[                    GELU] =  10.747 ms
  perf_total_per_op_us[                    NORM] =   3.074 ms
  perf_total_per_op_us[                 MUL_MAT] =   9.257 ms
  perf_total_per_op_us[                   SCALE] =   1.172 ms
  perf_total_per_op_us[                     CPY] =   5.760 ms
  perf_total_per_op_us[                 RESHAPE] =   0.105 ms
  perf_total_per_op_us[                    VIEW] =   0.739 ms
  perf_total_per_op_us[                 PERMUTE] =   0.318 ms
  perf_total_per_op_us[               TRANSPOSE] =   0.104 ms
  perf_total_per_op_us[                GET_ROWS] =   0.028 ms
  perf_total_per_op_us[           DIAG_MASK_INF] =   0.731 ms
  perf_total_per_op_us[                SOFT_MAX] =   1.544 ms
  perf_total_per_op_us[                   ALIBI] =   7.019 ms
  perf_total_per_op_us[           INNER PRODUCT] =  28.004 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.318 ms
  perf_total_per_op_us[                     MUL] =   0.399 ms
  perf_total_per_op_us[                    GELU] =  10.749 ms
  perf_total_per_op_us[                    NORM] =   3.140 ms
  perf_total_per_op_us[                 MUL_MAT] =   9.072 ms
  perf_total_per_op_us[                   SCALE] =   1.109 ms
  perf_total_per_op_us[                     CPY] =   5.829 ms
  perf_total_per_op_us[                 RESHAPE] =   0.104 ms
  perf_total_per_op_us[                    VIEW] =   0.816 ms
  perf_total_per_op_us[                 PERMUTE] =   0.322 ms
  perf_total_per_op_us[               TRANSPOSE] =   0.105 ms
  perf_total_per_op_us[                GET_ROWS] =   0.029 ms
  perf_total_per_op_us[           DIAG_MASK_INF] =   0.716 ms
  perf_total_per_op_us[                SOFT_MAX] =   1.591 ms
  perf_total_per_op_us[                   ALIBI] =   7.062 ms
  perf_total_per_op_us[           INNER PRODUCT] =  27.667 ms
  ========================================

  model_print_timings:        load time = 13995.45 ms
  model_print_timings:      sample time =     2.73 ms /     3 runs   (    0.91 ms per token)
  model_print_timings: prompt eval time = 12656.38 ms /   988 tokens (   12.81 ms per token)
  model_print_timings:        eval time =   146.90 ms /     2 runs   (   73.45 ms per token)
  model_print_timings:       total time = 14196.92 ms
  ========== eval time log of each prediction ==========
  prediction   0, time: 12656.38ms
  prediction   1, time: 73.60ms
  prediction   2, time: 73.29ms
  ```
  </details>
- <details>
  <summary>
  graph-mpt-mha-support <code>ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 48 -b 2048 -c 2048 -n 3 --memory-auto -p "$(echo <a href="https://pastebin.com/Qgs4X0VC">"$LUOYU_PROMPT"</a> | cut -d' ' -f 1-800)"</code>
  </summary>

  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =  23.430 ms
  perf_total_per_op_us[                     MUL] =  25.247 ms
  perf_total_per_op_us[                  REPEAT] = 130.404 ms
  perf_total_per_op_us[                    GELU] = 208.690 ms
  perf_total_per_op_us[                    NORM] =  27.116 ms
  perf_total_per_op_us[                 MUL_MAT] = 901.936 ms
  perf_total_per_op_us[                    VIEW] =   1.185 ms
  perf_total_per_op_us[                 PERMUTE] =   0.100 ms
  perf_total_per_op_us[                GET_ROWS] =   5.114 ms
  perf_total_per_op_us[              FLASH_ATTN] = 105.717 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =  61.095 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.403 ms
  perf_total_per_op_us[                     MUL] =   0.414 ms
  perf_total_per_op_us[                    GELU] =  10.541 ms
  perf_total_per_op_us[                    NORM] =   2.934 ms
  perf_total_per_op_us[                    VIEW] =   1.204 ms
  perf_total_per_op_us[                 PERMUTE] =   0.101 ms
  perf_total_per_op_us[                GET_ROWS] =   0.028 ms
  perf_total_per_op_us[              FLASH_ATTN] =   4.067 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =   2.320 ms
  perf_total_per_op_us[           INNER PRODUCT] =  26.553 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.317 ms
  perf_total_per_op_us[                     MUL] =   0.401 ms
  perf_total_per_op_us[                    GELU] =  10.622 ms
  perf_total_per_op_us[                    NORM] =   2.841 ms
  perf_total_per_op_us[                    VIEW] =   1.298 ms
  perf_total_per_op_us[                 PERMUTE] =   0.148 ms
  perf_total_per_op_us[                GET_ROWS] =   0.014 ms
  perf_total_per_op_us[              FLASH_ATTN] =   4.145 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =   2.270 ms
  perf_total_per_op_us[           INNER PRODUCT] =  26.537 ms
  ========================================

  model_print_timings:        load time =  2834.27 ms
  model_print_timings:      sample time =     2.64 ms /     3 runs   (    0.88 ms per token)
  model_print_timings: prompt eval time =  1494.47 ms /   988 tokens (    1.51 ms per token)
  model_print_timings:        eval time =   103.54 ms /     2 runs   (   51.77 ms per token)
  model_print_timings:       total time =  2992.35 ms
  ========== eval time log of each prediction ==========
  prediction   0, time: 1494.47ms
  prediction   1, time: 51.76ms
  prediction   2, time: 51.79ms
  ```
  </details>
- <details>
  <summary>
  graph-mpt-mha-support <code>ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 3 --memory-auto -p "$(echo <a href="https://pastebin.com/Qgs4X0VC">"$LUOYU_PROMPT"</a> | cut -d' ' -f 1-800)"</code>
  </summary>

  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =  23.198 ms
  perf_total_per_op_us[                     MUL] =  22.576 ms
  perf_total_per_op_us[                  REPEAT] = 137.497 ms
  perf_total_per_op_us[                    GELU] = 188.130 ms
  perf_total_per_op_us[                    NORM] =  27.938 ms
  perf_total_per_op_us[                 MUL_MAT] = 761.702 ms
  perf_total_per_op_us[                    VIEW] =   1.222 ms
  perf_total_per_op_us[                 PERMUTE] =   0.103 ms
  perf_total_per_op_us[                GET_ROWS] =   5.423 ms
  perf_total_per_op_us[              FLASH_ATTN] =  99.823 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =  58.736 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.335 ms
  perf_total_per_op_us[                     MUL] =   0.418 ms
  perf_total_per_op_us[                    GELU] =  10.718 ms
  perf_total_per_op_us[                    NORM] =   2.877 ms
  perf_total_per_op_us[                    VIEW] =   1.148 ms
  perf_total_per_op_us[                 PERMUTE] =   0.106 ms
  perf_total_per_op_us[                GET_ROWS] =   0.015 ms
  perf_total_per_op_us[              FLASH_ATTN] =   4.181 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =   2.287 ms
  perf_total_per_op_us[           INNER PRODUCT] =  26.710 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.354 ms
  perf_total_per_op_us[                     MUL] =   0.398 ms
  perf_total_per_op_us[                    GELU] =  10.848 ms
  perf_total_per_op_us[                    NORM] =   3.154 ms
  perf_total_per_op_us[                    VIEW] =   1.232 ms
  perf_total_per_op_us[                 PERMUTE] =   0.106 ms
  perf_total_per_op_us[                GET_ROWS] =   0.027 ms
  perf_total_per_op_us[              FLASH_ATTN] =   4.034 ms
  perf_total_per_op_us[    FLASH_ATTN_KV_UPDATE] =   2.443 ms
  perf_total_per_op_us[           INNER PRODUCT] =  26.731 ms
  ========================================

  model_print_timings:        load time =  2700.29 ms
  model_print_timings:      sample time =     2.79 ms /     3 runs   (    0.93 ms per token)
  model_print_timings: prompt eval time =  1330.82 ms /   988 tokens (    1.35 ms per token)
  model_print_timings:        eval time =   104.65 ms /     2 runs   (   52.33 ms per token)
  model_print_timings:       total time =  2859.34 ms
  ========== eval time log of each prediction ==========
  prediction   0, time: 1330.82ms
  prediction   1, time: 52.03ms
  prediction   2, time: 52.63ms
  ```
  </details>
- <details>
  <summary>
  graph-mpt-mha-support <code>ENGINE_PROFILING=1 numactl -m 1 -C 56-111 bin/run_mpt -m /home_local/dingyi/data/mpt-7b-pr447-q4j-sym-int8-fp32-g128.bin --seed 1234 -t 56 -b 2048 -c 2048 -n 3 --memory-f16 -p "$(echo <a href="https://pastebin.com/Qgs4X0VC">"$LUOYU_PROMPT"</a> | cut -d' ' -f 1-800)"</code>
  </summary>

  ```
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =  22.522 ms
  perf_total_per_op_us[                     MUL] =  23.008 ms
  perf_total_per_op_us[                  REPEAT] = 128.805 ms
  perf_total_per_op_us[                    GELU] = 188.478 ms
  perf_total_per_op_us[                    NORM] =  25.166 ms
  perf_total_per_op_us[                 MUL_MAT] = 2411.392 ms
  perf_total_per_op_us[                   SCALE] = 1052.907 ms
  perf_total_per_op_us[                     CPY] = 140.443 ms
  perf_total_per_op_us[                 RESHAPE] =   0.107 ms
  perf_total_per_op_us[                    VIEW] =   0.765 ms
  perf_total_per_op_us[                 PERMUTE] =   0.321 ms
  perf_total_per_op_us[               TRANSPOSE] =   0.132 ms
  perf_total_per_op_us[                GET_ROWS] =   4.424 ms
  perf_total_per_op_us[           DIAG_MASK_INF] = 363.926 ms
  perf_total_per_op_us[                SOFT_MAX] =  73.186 ms
  perf_total_per_op_us[                   ALIBI] = 7920.785 ms
  perf_total_per_op_us[           INNER PRODUCT] =   0.000 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.357 ms
  perf_total_per_op_us[                     MUL] =   0.369 ms
  perf_total_per_op_us[                    GELU] =  10.764 ms
  perf_total_per_op_us[                    NORM] =   2.420 ms
  perf_total_per_op_us[                 MUL_MAT] =   8.110 ms
  perf_total_per_op_us[                   SCALE] =   1.234 ms
  perf_total_per_op_us[                     CPY] =   4.041 ms
  perf_total_per_op_us[                 RESHAPE] =   0.114 ms
  perf_total_per_op_us[                    VIEW] =   0.757 ms
  perf_total_per_op_us[                 PERMUTE] =   0.333 ms
  perf_total_per_op_us[               TRANSPOSE] =   0.111 ms
  perf_total_per_op_us[                GET_ROWS] =   0.021 ms
  perf_total_per_op_us[           DIAG_MASK_INF] =   0.686 ms
  perf_total_per_op_us[                SOFT_MAX] =   1.152 ms
  perf_total_per_op_us[                   ALIBI] =   7.307 ms
  perf_total_per_op_us[           INNER PRODUCT] =  25.309 ms
  ========================================
  === GRAPH Profiling ===
  perf_total_per_op_us[                     ADD] =   0.344 ms
  perf_total_per_op_us[                     MUL] =   0.363 ms
  perf_total_per_op_us[                    GELU] =  10.200 ms
  perf_total_per_op_us[                    NORM] =   2.411 ms
  perf_total_per_op_us[                 MUL_MAT] =   8.056 ms
  perf_total_per_op_us[                   SCALE] =   1.046 ms
  perf_total_per_op_us[                     CPY] =   4.097 ms
  perf_total_per_op_us[                 RESHAPE] =   0.103 ms
  perf_total_per_op_us[                    VIEW] =   0.754 ms
  perf_total_per_op_us[                 PERMUTE] =   0.309 ms
  perf_total_per_op_us[               TRANSPOSE] =   0.105 ms
  perf_total_per_op_us[                GET_ROWS] =   0.021 ms
  perf_total_per_op_us[           DIAG_MASK_INF] =   0.705 ms
  perf_total_per_op_us[                SOFT_MAX] =   1.152 ms
  perf_total_per_op_us[                   ALIBI] =   7.020 ms
  perf_total_per_op_us[           INNER PRODUCT] =  25.661 ms
  ========================================

  model_print_timings:        load time = 13633.52 ms
  model_print_timings:      sample time =     2.68 ms /     3 runs   (    0.89 ms per token)
  model_print_timings: prompt eval time = 12361.66 ms /   988 tokens (   12.51 ms per token)
  model_print_timings:        eval time =   133.83 ms /     2 runs   (   66.92 ms per token)
  model_print_timings:       total time = 13821.57 ms
  ========== eval time log of each prediction ==========
  prediction   0, time: 12361.66ms
  prediction   1, time: 67.24ms
  prediction   2, time: 66.59ms
  ```
  </details>
</details>

## Dependency Change?
N/A

### CODE REVIEW

**Review by zhenwei-intel — APPROVED** [review]

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp [hunk 1]
```diff
@@ -961,31 +961,46 @@ class ScaleTrackMax<ISA_T, float, float> {
...
+
+  template <bool HAS_ALIBI>
+  JBLAS_CODE forward_(const SType* src, const int src_step, const int M_offset, const int N_offset, const int M,
+                      const int N, const Param& p) const {
     const auto dst = p.dst + M_offset * p.ld_dst + N_offset;
     const auto dst_max = p.dst_max + M_offset;
 #if CompileAVX512F()
 #if MHA_2ND_EXP
     const auto v_scale = _mm512_set1_ps(p.scale);
+    const auto v_seq15 = _mm512_loadu_ps(seq15);
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:980]: I wonder whether these will infect performance if ```forward<false>```, will compiler allocate registers for these?

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp [hunk 2]
```diff
@@ -961,31 +961,46 @@ class ScaleTrackMax<ISA_T, float, float> {
...
+      auto alibi_curr = alibi_base;
       const auto N_unmasked =
           std::min(N, p.causal_offset < 0 ? INT32_MAX : i + M_offset - N_offset + p.causal_offset + 1);
 
       const auto v_mask = _cvtu32_mask16((1U << (N_unmasked % 16)) - 1);
       int j = 0;
       auto v_max = _mm512_set1_ps(-INFINITY);
       for (; j < N_unmasked - 15; j += 16) {
-        const auto xs = _mm512_mul_ps(v_scale, _mm512_loadu_ps(src + i * src_step + j));
+        const auto xs = _mm512_fmadd_ps(v_scale, _mm512_loadu_ps(src + i * src_step + j), alibi_curr);
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:994]: can we avoid it when ```forward<false>```

Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:994]: I think fmadd should not have extra overhead over mul.

File: intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp (hunk 1 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/core/layers/mha_dense.cpp:980]: I don't think so. Everything except `alibi_curr` are not used in the innermost loop. In addition, local performance test && CI shows that there is no performance regression on models without AliBi (e.g. llama).

**Review by airMeng — APPROVED** [review]
take this as an example for MHA introduction

**Pull request merged**